Gems/Atom/Feature/Common/Assets/ShaderLib/Atom/Features/Shadow/DirectionalLightShadowCalculator.azsli

```diff
@@ -0,0 +1,416 @@
+        case ShadowFilterMethod::Pcf:
+            lit = GetVisibilityFromLightPcf();
+            break;
+        case ShadowFilterMethod::Esm:
+            lit = GetVisibilityFromLightEsm();
+            break;
+        case ShadowFilterMethod::EsmPcf:
+            lit = GetVisibilityFromLightEsmPcf();
+            break;
+    }
+                
+    return lit;
+}
+
+float DirectionalShadowCalculator::GetVisibilityFromLightNoFilter()
+{       
+    const uint cascadeCount = ViewSrg::m_directionalLightShadows[m_lightIndex].m_cascadeCount;
+    for (uint indexOfCascade = 0; indexOfCascade < cascadeCount; ++indexOfCascade)
+    {
+        const bool2 checkedShadowed = IsShadowed(
+            m_shadowCoords[indexOfCascade],
+            indexOfCascade);
+        if (checkedShadowed.x)
+        {
+            m_debugInfo.m_cascadeIndex = indexOfCascade;
+            return checkedShadowed.y ? 0. : 1.;
+        }
+    }
+    m_debugInfo.m_cascadeIndex = cascadeCount;
+    return 1.;
+}    
+
+float DirectionalShadowCalculator::SamplePcfBicubic(const float3 shadowCoord, const uint indexOfCascade)
+{
+    const uint filteringSampleCount = ViewSrg::m_directionalLightShadows[m_lightIndex].m_filteringSampleCount;
+    const uint size = ViewSrg::m_directionalLightShadows[m_lightIndex].m_shadowmapSize;
+    
+    const uint cascadeCount = ViewSrg::m_directionalLightShadows[m_lightIndex].m_cascadeCount;
+
+    SampleShadowMapBicubicParameters param;
+    param.shadowMap = m_directionalLightShadowmap;
+    param.shadowPos = float3(shadowCoord.xy, indexOfCascade);
+    param.shadowMapSize = size;
+    param.invShadowMapSize = rcp(size); 
+    param.comparisonValue = shadowCoord.z;
+    param.samplerState = SceneSrg::m_hwPcfSampler;
+    param.receiverPlaneDepthBias = m_receiverShadowPlaneBiasEnable ? ComputeReceiverPlaneDepthBias(m_shadowPosDX[indexOfCascade], m_shadowPosDY[indexOfCascade]) : 0;
+    
+    [branch]     
+    if (filteringSampleCount <= 4)       
+    {
+        return SampleShadowMapBicubic_4Tap(param);
+    }
+    else if (filteringSampleCount <= 9)
+    {
+        return SampleShadowMapBicubic_9Tap(param);
+    }
+    else
+    {
+        return SampleShadowMapBicubic_16Tap(param);
+    }
+}
+
+float DirectionalShadowCalculator::GetVisibilityFromLightPcf()
+{   
+    static const float PixelMargin = 1.5; // avoiding artifact between cascade levels.
+    static const float DepthMargin = 1e-8; // avoiding artifact when near depth bounds.
+
+    bool cascadeFound = false;
+    int currentCascadeIndex = 0;
+
+    const uint size = ViewSrg::m_directionalLightShadows[m_lightIndex].m_shadowmapSize;
+    const uint cascadeCount = ViewSrg::m_directionalLightShadows[m_lightIndex].m_cascadeCount;
+    for (uint indexOfCascade = 0; indexOfCascade < cascadeCount; ++indexOfCascade)
+    {
+        const float3 shadowCoord = m_shadowCoords[indexOfCascade];
+        
+        if (shadowCoord.x >= 0. && shadowCoord.x * size < size - PixelMargin &&
```

Comment:
After seeing this condition a number of times I wonder if it's worth refactoring it to be a helper member function